crates/starknet-os/src/execution/syscall_handler.rs

```diff
@@ -529,44 +533,80 @@ impl SyscallHandler for StorageWriteHandler {
     }
 }
 
-// TODO: Keccak syscall.
-// #[derive(Debug, Eq, PartialEq)]
-// pub struct KeccakRequest {
-//     pub input_start: Relocatable,
-//     pub input_end: Relocatable,
-// }
-//
-// impl SyscallRequest for KeccakRequest {
-//     fn read(vm: &VirtualMachine, ptr: &mut Relocatable) -> SyscallResult<KeccakRequest> {
-//         let input_start = vm.get_relocatable(*ptr)?;
-//         *ptr = (*ptr + 1)?;
-//         let input_end = vm.get_relocatable(*ptr)?;
-//         *ptr = (*ptr + 1)?;
-//         Ok(KeccakRequest { input_start, input_end })
-//     }
-// }
-//
-// #[derive(Debug, Eq, PartialEq)]
-// pub struct KeccakResponse {
-//     pub result_low: Felt252,
-//     pub result_high: Felt252,
-// }
-//
-// impl SyscallResponse for KeccakResponse {
-//     fn write(self, vm: &mut VirtualMachine, ptr: &mut Relocatable) -> WriteResponseResult {
-//         write_felt(vm, ptr, self.result_low)?;
-//         write_felt(vm, ptr, self.result_high)?;
-//         Ok(())
-//     }
-// }
-//
-// pub fn keccak(
-//     request: KeccakRequest,
-//     vm: &mut VirtualMachine,
-//     syscall_handler: &mut SyscallHintProcessor<'_>,
-//     remaining_gas: &mut u64,
-// ) -> SyscallResult<KeccakResponse> {
-// }
+pub struct KeccakHandler;
+pub struct KeccakRequest {
+    pub input_start: Relocatable,
+    pub input_end: Relocatable,
+}
+pub struct KeccakResponse {
+    pub result_low: Felt252,
+    pub result_high: Felt252,
+}
+
+impl SyscallHandler for KeccakHandler {
+    type Request = KeccakRequest;
+    type Response = KeccakResponse;
+
+    fn read_request(vm: &VirtualMachine, ptr: &mut Relocatable) -> SyscallResult<Self::Request> {
+        let input_start = vm.get_relocatable(*ptr)?;
+        *ptr = (*ptr + 1)?;
+        let input_end = vm.get_relocatable(*ptr)?;
+        *ptr = (*ptr + 1)?;
+        Ok(KeccakRequest { input_start, input_end })
+    }
+
+    async fn execute<S>(
+        request: Self::Request,
+        vm: &mut VirtualMachine,
+        _exec_wrapper: &mut ExecutionHelperWrapper<S>,
+        remaining_gas: &mut u64,
+    ) -> SyscallResult<Self::Response>
+    where
+        S: Storage + 'static,
+    {
+        let input_len = (request.input_end - request.input_start)?;
+        // The to_usize unwrap will not fail as the constant value is 17
+        let (n_rounds, remainder) = num_integer::div_rem(input_len, KECCAK_FULL_RATE_IN_U64S.to_usize().unwrap());
+
+        if remainder != 0 {
+            return Err(SyscallExecutionError::SyscallError {
+                error_data: vec![Felt252::from_hex_unchecked(INVALID_INPUT_LENGTH_ERROR)],
+            });
+        }
+        let gas_cost = n_rounds.to_u64().unwrap() * KECCAK_ROUND_COST_GAS_COST;
+        if gas_cost > *remaining_gas {
+            return Err(SyscallExecutionError::OutOfGas { remaining_gas: (*remaining_gas) });
+        }
+        *remaining_gas -= gas_cost;
```

Comment:
This should be checked math to ensure that it doesn't underflow, although I suspect that an underflow would cause some other failure...

Reply:
@notlesh how this can underflow if im checking `if gas_cost > *remaining_gas`?

Reply:
Oh, you're right. Sorry, I've been conditioned to avoid the arithmetic operators altogether in favor of something like this: https://docs.rs/num/latest/num/trait.CheckedSub.html

Your check is sufficient :+1: